### PR TITLE
kosaraju algorithm.cpp

### PR DESCRIPTION
**Pull request by pragyajaiswa05l**
To find strongly connected graph

Fixes #2246 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added code for Kosaraju Algorithm in c++
- 

#### Languages Used:
- c++
- 

#### Files Added:
- kosaraju algorithm.cpp
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by raghav-dalmia** [comment]
save file with .cpp extension.

**Comment by pragyajaiswa05l** [comment]
done please review it

**Comment by raghav-dalmia** [comment]
I didn't tell you to change the PR title. The file which is in this PR doesn't have any extension. Save it with .cpp and then update that file here. Also, mention the issue number in the PR description.

**Comment by pragyajaiswa05l** [comment]
done please review it @raghav-dalmia

**Comment by raghav-dalmia** [comment]
still, your file doesn't have any extension. Review it yourself.

**Comment by pragyajaiswa05l** [comment]
file added should be named as   - kosaraju algorithm .cpp ? @raghav-dalmia

**Comment by pragyajaiswa05l** [comment]
please help me how to do it .i m new so not able to correct it @raghav-dalmia

**Comment by Vaibhavraj10** [comment]
> please help me how to do it .i m new so not able to correct it @raghav-dalmia

Go into file changed tab. Click on edit this file. And change the file name. Add the file extention to the file from kosaraju algorithm to kosaraju_algorithm.cpp

**Comment by pragyajaiswa05l** [comment]
@Vaibhavraj10  @raghav-dalmia  @jainaman224   done please review it

File: kosaraju_Algorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
+//Strongly connected components(KOSARAJU ALGORITHM)
+#include <iostream>
+#include <vector>
+#include <stack>
+#include <unordered_set>
+using namespace std;
+
+void dfs(vector<int>* edges, int start, unordered_set<int> &visited, stack<int> &finishStack) { //dfs to store vertices in a stack as per their finish time
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: take the comment into the previous line

File: kosaraju_Algorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	visited.insert(start);
+	for (int i = 0; i < edges[start].size(); i++) {
+		int adjacent = edges[start][i];
+		if (visited.count(adjacent) == 0) {
+			dfs(edges, adjacent, visited, finishStack);
+		}
+	}
+	finishStack.push(start);
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: correct the indentation.keep four spaces instead of tab

File: kosaraju_Algorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+3 5
+5 6
+6 5
+
+OUTPUT
+6 5
+2 3 4 1         */
+
+
+
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: remove this lines

Comment (pragyajaiswa05l) [reply kosaraju_Algorithm.cpp]: @asha15 done please review it

File: kosaraju_Algorithm.cpp (hunk 2 above)
Comment (pragyajaiswa05l) [reply kosaraju_Algorithm.cpp]: @asha15 done  please review it

**Comment by pragyajaiswa05l** [comment]
@raghav-dalmia  @jainaman224  @jainaman224  please review it i have the changes which were required

File: kosaraju_Algorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+void dfs2(vector<int>* edges, int start, unordered_set<int>* component, unordered_set<int> & visited) { //dfs of a transpose graph
+	visited.insert(start);
+	component->insert(start);
+	for (int i = 0; i < edges[start].size(); i++) {
+		int adjacent = edges[start][i];
+		if (visited.count(adjacent) == 0) {
+			dfs2(edges, adjacent, component, visited);
+		}
+	}
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: keep four spaces instead of tab

File: kosaraju_Algorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+}
+int main() {
+    int n;                                             //no.of vertices
+    cout<<"Enter the no. of vertices : "<<endl;
+    cin >> n;
+    vector<int>* edges = new vector<int>[n];           //using a adjacency list to implement
+    vector<int>* edgesT = new vector<int>[n];         //transpose of a graph
+    int m;                                             // no.of edges
+    cout<<"Enter the no. of edges : "<<endl;
+    cin >> m;
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: keep a line space

File: kosaraju_Algorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    for (int i = 0; i < m; i++) {
+        int j, k;
+        cin >> j >> k;
+        edges[j - 1].push_back(k - 1);
+        edgesT[k - 1].push_back(j - 1);
+    }
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: keep a line space after this

File: kosaraju_Algorithm.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    unordered_set<unordered_set<int>*>* components = getSCC(edges, edgesT, n);
+    unordered_set<unordered_set<int>*>::iterator it = components->begin();
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: keep a line space after this

File: kosaraju_Algorithm.cpp (hunk 5 above)
Comment (pragyajaiswa05l) [reply kosaraju_Algorithm.cpp]: @asha15  i didn't get u can u please say what u want to say by "keep a line space"

**Comment by pragyajaiswa05l** [comment]
@asha15 @raghav-dalmia  @jainaman224  please review it  i have made a changes which were required

**Comment by raghav-dalmia** [comment]
Squash your commits. Refer [this](https://github.com/wprig/wprig/wiki/How-to-squash-commits).

File: kosaraju_Algorithm.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+void dfs(vector<int>* edges, int start, unordered_set<int> &visited, stack<int> &finishStack) { 
+    visited.insert(start);
+    for (int i = 0; i < edges[start].size(); i++) {
+        int adjacent = edges[start][i];
+        if (visited.count(adjacent) == 0) {
+            dfs(edges, adjacent, visited, finishStack);
+        }
+    }
+    finishStack.push(start);
+}
```
Comment (asha15) [inline kosaraju_Algorithm.cpp]: keep a line space after this

**Comment by pragyajaiswa05l** [comment]
@asha15  how to squash it please help. New to github doesnot know how to do it

File: kosaraju_Algorithm.cpp (hunk 5 above)
Comment (asha15) [reply kosaraju_Algorithm.cpp]: keep a blank line after this line

**Comment by asha15** [comment]
Please reach me on slack

**Comment by pragyajaiswa05l** [comment]
yes mentor @asha15

**Comment by asha15** [comment]
include this file inside the correct folder. then we are good to merge :)

**Comment by MastersAbh** [comment]
@pragyajaiswa05l are you working on this ?

**Comment by pragyajaiswa05l** [comment]
I have completed it

File: kosaraju_Algorithm.cpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
+//Strongly connected components(KOSARAJU ALGORITHM)
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:1]: Change the file name to Kosaraju_Algorithm/Kosaraju_Algorithm.cpp

File: kosaraju_Algorithm.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    }
+	
+    unordered_set<unordered_set<int>*>* components = getSCC(edges, edgesT, n);
+    unordered_set<unordered_set<int>*>::iterator it = components->begin();
+	
+    while (it != components->end()) {
+        unordered_set<int>* component = *it;
+        unordered_set<int>::iterator it2 = component->begin();
+        while (it2 != component->end()) {
+            cout <<*it2  + 1<< " ";
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:79]: add space around << and >>

File: kosaraju_Algorithm.cpp [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        edgesT[k - 1].push_back(j - 1);
+    }
+	
+    unordered_set<unordered_set<int>*>* components = getSCC(edges, edgesT, n);
+    unordered_set<unordered_set<int>*>::iterator it = components->begin();
+	
+    while (it != components->end()) {
+        unordered_set<int>* component = *it;
+        unordered_set<int>::iterator it2 = component->begin();
+        while (it2 != component->end()) {
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:78]: add space around ->

File: kosaraju_Algorithm.cpp [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        edges[j - 1].push_back(k - 1);
+        edgesT[k - 1].push_back(j - 1);
+    }
+	
+    unordered_set<unordered_set<int>*>* components = getSCC(edges, edgesT, n);
+    unordered_set<unordered_set<int>*>::iterator it = components->begin();
+	
+    while (it != components->end()) {
+        unordered_set<int>* component = *it;
+        unordered_set<int>::iterator it2 = component->begin();
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:77]: add space around ->

File: kosaraju_Algorithm.cpp [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        int j, k;
+        cin >> j >> k;
+        edges[j - 1].push_back(k - 1);
+        edgesT[k - 1].push_back(j - 1);
+    }
+	
+    unordered_set<unordered_set<int>*>* components = getSCC(edges, edgesT, n);
+    unordered_set<unordered_set<int>*>::iterator it = components->begin();
+	
+    while (it != components->end()) {
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:75]: add space around ->

File: kosaraju_Algorithm.cpp [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+int main() {
+    int n;                                                 //no.of vertices
+    cout<<"Enter the no. of vertices : "<<endl;
+    cin >> n;
+    vector<int>* edges = new vector<int>[n];   //using a adjacency list to implement
+    vector<int>* edgesT = new vector<int>[n];             //transpose of a graph
+    int m;                                                // no.of edges
+    cout<<"Enter the no. of edges : "<<endl;
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:62]: add space around <<

File: kosaraju_Algorithm.cpp [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        unordered_set<int>* component = new unordered_set<int>();
+        dfs2(edgesT, element, component, visited);
+        output->insert(component);
+    }
+    return output;
+}
+
+int main() {
+    int n;                                                 //no.of vertices
+    cout<<"Enter the no. of vertices : "<<endl;
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:57]: add space around <<

File: kosaraju_Algorithm.cpp [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    visited.clear();
+    while (finishedVertices.size() != 0) {
+        int element = finishedVertices.top();
+        finishedVertices.pop();
+        if (visited.count(element) != 0) {
+            continue;
+        }
+        unordered_set<int>* component = new unordered_set<int>();
+        dfs2(edgesT, element, component, visited);
+        output->insert(component);
```
Comment (somya-kapoor) [inline kosaraju_Algorithm.cpp:50]: add space around ->

**Comment by somya-kapoor** [comment]
Are you working on it?

**Comment by somya-kapoor** [comment]
Due to no activity for more than 7 days, I am closing this PR.

**Pull request closed without merging**